src/Storages/RocksDB/EmbeddedRocksDBBulkSink.cpp

```diff
@@ -0,0 +1,235 @@
+#include <atomic>
+#include <cstddef>
+#include <cstdlib>
+#include <filesystem>
+#include <memory>
+#include <optional>
+#include <random>
+#include <stdatomic.h>
+#include <IO/WriteBufferFromString.h>
+#include <Storages/RocksDB/EmbeddedRocksDBBulkSink.h>
+#include <Storages/RocksDB/StorageEmbeddedRocksDB.h>
+
+#include <Columns/ColumnString.h>
+#include <Core/SortDescription.h>
+#include <DataTypes/DataTypeString.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/Context.h>
+#include <rocksdb/options.h>
+#include <rocksdb/status.h>
+#include <rocksdb/utilities/db_ttl.h>
+#include <Common/SipHash.h>
+#include <Common/getRandomASCIIString.h>
+#include <Common/CurrentThread.h>
+#include <Common/MemoryTrackerBlockerInThread.h>
+#include <Common/logger_useful.h>
+#include <Common/scope_guard_safe.h>
+#include <Common/setThreadName.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ROCKSDB_ERROR;
+}
+
+static const IColumn::Permutation & getAscendingPermutation(const IColumn & column, IColumn::Permutation & perm)
+{
+    column.getPermutation(IColumn::PermutationSortDirection::Ascending, IColumn::PermutationSortStability::Stable, 0, 1, perm);
+    return perm;
+}
+
+/// Build SST file from key-value pairs
+static rocksdb::Status buildSSTFile(const String & path, const ColumnString & keys, const ColumnString & values, const std::optional<IColumn::Permutation> & perm_ = {})
+{
+    /// rocksdb::SstFileWriter requires keys to be sorted in ascending order
+    IColumn::Permutation calculated_perm;
+    const IColumn::Permutation & perm = perm_ ? *perm_ : getAscendingPermutation(keys, calculated_perm);
+
+    rocksdb::SstFileWriter sst_file_writer(rocksdb::EnvOptions{}, rocksdb::Options{});
+    auto status = sst_file_writer.Open(path);
+    if (!status.ok())
+        return status;
+
+    auto rows = perm.size();
+    WriteBufferFromOwnString wb_value;
+    for (size_t i = 0; i < rows; ++i)
+    {
+        auto row = perm[i];
+
+        status = sst_file_writer.Put(keys.getDataAt(row).toView(), values.getDataAt(row).toView());
+
+        /// There could be duplicated keys in chunk, thus Put may give IsInvalidArgument. This is ok, as we're certain that
```

Comment:
It's about performance. `Put` will compare keys before insertion, so I think we don't need to do it before `Put`. Regarding concern about `IsInvalidArgument`, can we compared the error message and "Keys must be added in strict ascending order.". Wdyt?
```cpp
    if (file_info.num_entries == 0) {
      file_info.smallest_key.assign(user_key.data(), user_key.size());
    } else {
      if (internal_comparator.user_comparator()->Compare(
              user_key, file_info.largest_key) <= 0) {
        // Make sure that keys are added in order
        return Status::InvalidArgument(
            "Keys must be added in strict ascending order.");
      }
    }
```

Reply:
It's about performance. `Put` already compares keys before insertion, so I think we don't need to do it here.

```cpp
    if (file_info.num_entries == 0) {
      file_info.smallest_key.assign(user_key.data(), user_key.size());
    } else {
      if (internal_comparator.user_comparator()->Compare(
              user_key, file_info.largest_key) <= 0) {
        // Make sure that keys are added in order
        return Status::InvalidArgument(
            "Keys must be added in strict ascending order.");
      }
```

Regarding concern about `IsInvalidArgument`, I think if status code is `IsInvalidArgument` then we also compare the status message to  "Keys must be added in strict ascending order." to verify it's indeed indicate a duplicated key. Do you think this approach is good?

Reply:
I changed the code to deduplicate keys before `Put`. The previous implementation is not correct because we suppose to select the last value of the if key is duplicated.